source/index.ts

```diff
@@ -1,47 +1,130 @@
 import * as util from 'util';
 
+export enum TypeName {
+	null = 'null',
+	boolean = 'boolean',
+	undefined = 'undefined',
+	string = 'string',
+	number = 'number',
+	symbol = 'symbol',
+	Function = 'Function',
+	Array = 'Array',
+	Buffer = 'Buffer',
+	Object = 'Object',
+	RegExp = 'RegExp',
+	Date = 'Date',
+	Error = 'Error',
+	Map = 'Map',
+	Set = 'Set',
+	WeakMap = 'WeakMap',
+	WeakSet = 'WeakSet',
+	Int8Array = 'Int8Array',
+	Uint8Array = 'Uint8Array',
+	Uint8ClampedArray = 'Uint8ClampedArray',
+	Int16Array = 'Int16Array',
+	Uint16Array = 'Uint16Array',
+	Int32Array = 'Int32Array',
+	Uint32Array = 'Uint32Array',
+	Float32Array = 'Float32Array',
+	Float64Array = 'Float64Array',
+	ArrayBuffer = 'ArrayBuffer',
+	SharedArrayBuffer = 'SharedArrayBuffer',
+	Promise = 'Promise'
+}
+
 const toString = Object.prototype.toString;
-const getObjectType = (value: any) => toString.call(value).slice(8, -1) as string;
 const isOfType = (type: string) => (value: any) => typeof value === type; // tslint:disable-line:strict-type-predicates
-const isObjectOfType = (type: string) => (value: any) => getObjectType(value) === type;
 
-function is(value: any) { // tslint:disable-line:only-arrow-functions
+const getObjectType = (value: any): TypeName | null => {
+	const objectName = toString.call(value).slice(8, -1) as string;
+
+	switch (objectName) {
+		case 'ArrayBuffer':
+			return TypeName.ArrayBuffer;
```

Comment:
I don't understand this `switch`. If the value is `"ArrayBuffer"`, you are returning `"ArrayBuffer"`. The only purpose of it's seems to be to get TypeScript to accept that it's a `TypeName`, but that's not very idiomatic for TypeScript. Idiomatic is to not have anything in the runtime JS for type reasons - if you only want to work around the type system, use a cast and be done with it.

Reply:
Could you point to a normative source that explains how this approach is not idiomatic / considered a bad practice?

Reply:
Yeah, you can extract it from the design goals of the language: https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals

> # Goals

> 3. Impose no runtime overhead on emitted programs.

> 4. Emit **clean**, idiomatic, recognizable JavaScript code.

> 9. Use a consistent, **fully erasable**, structural type system.

> # Non-Goals

> 5. Add or rely on run-time type information in programs, or emit different code based on the results of the type system. Instead, encourage programming patterns that do not require run-time metadata.

Basically, converting a codebase from JS to TS should not change the JavaScript at all. The typings are a build-time construct that gets completely removed and should result in the same JavaScript as before.